### ISSUE TITLE
Global configuration option to require explicit indexing for all axes compliant with array API standard

### ISSUE THREAD

**Issue by Gattocrucco** [jax-ml/jax#37735]
The array API standard states (https://data-apis.org/array-api/latest/API_specification/indexing.html#multi-axis-indexing):

> Except in the case of providing a single ellipsis (e.g., A[2:10, ...] or A[1:, ..., 2:5]), the number of provided single-axis indexing expressions (excluding None) should equal N. For example, if A has rank 2, a single-axis indexing expression should be explicitly provided for both axes (e.g., A[2:10, :]). An IndexError exception should be raised if the number of provided single-axis indexing expressions (excluding None) is less than N.

I would like a jax config option that enables this check. Example:

```python
from jax import config
from jax import numpy as jnp

config.update('jax_enable_explicit_multi_axis_indexing', True)

x = jnp.eye(5)
x[0, :]  # ok
x[0, ...]  # ok
x[0]  # IndexError
```

**Comment by jakevdp** [comment]
Thanks for the request – this is quite a big change to make. It's not just about adding that config – if we want that config to be meaningful, then we need to strictly adhere to it in all JAX source code (otherwise, you could not enable that config and expect any JAX code to execute without error).

This part of the spec explicitly disagrees with NumPy's indexing semantics, and I've personally never understood the reason for it.

Can you say more about why you want this mode, and what it would be useful for?

**Comment by Gattocrucco** [comment]
First, this is not a priority, just something that would be nice, if it's complex I'd say not worth it.

The reason I would like this is to make code more readable and to avoid some bugs caused by wrong indexing expressions.

I think indexing expressions are more readable when you can see all indices in the code. Is `x[0]` a scalar or the first row of a matrix? `x[0, :]` makes that clear at a glance. Or I often have code paths that work for multiple array dimensionalities, having `x[..., 0, :]` makes it immediately clear the code path is meant to handle that. It's nice because it's both semantic and informative, much better that a comment/type hint while also being shorter. Enforcing this requires to check at runtime, so adherence to the rule would just be a way to enforce the convention.

Then, about avoiding bugs: last time I remember something like this happening I had some expressions like `x[0]`, then upstream in the code we changed `x` to 2d (was 1d before), the correct expression would have been `x[:, 0]`, not `x[0] == x[0, :]`, and we got a bug.

**Comment by jakevdp** [comment]
I added this flag as an experiment in https://github.com/jax-ml/jax/pull/38049, and enabled strict indexing mode in JAX's test suite. We could land something like this, but we'd have to first fix all the strict indexing failures within JAX itself!
